mods/bucket/init.lua

```diff
@@ -152,6 +157,16 @@ minetest.register_craftitem("bucket:bucket_empty", {
 			minetest.add_node(pointed_thing.under, {name="air"})
 
 			return ItemStack(giving_back)
+		else
+			-- non-liquid nodes will have their on_punch triggered
+			minetest.node_punch(pointed_thing.under, node, user, pointed_thing)
+			-- We have to explicitly trigger the node_def on_punch too
+			-- minetest.node_punch only triggers minetest.registered_on_punchnodes
+			local node_def = minetest.registered_nodes[node.name]
+			if node_def and node_def.on_punch then
+				node_def.on_punch(pointed_thing.under, node, user, pointed_thing)
```

Comment:
For nodes without custom `on_punch` you just triggered `node_punch` (through `__index` onto node defaults) twice.